### PR TITLE
Recognize and transform redundant calls in MH.invokeExact MethodHandle chain

### PR DESCRIPTION
**Pull request by nbhuiyan**
This changeset includes the following: 
* recognize `java/lang/invoke/Invokers` methods that require handling in RecognizedCallTransformer
* Transform `Invokers.checkExactType` to a ZEROCHK
* Eliminate `Invokers.checkCustomized` calls when MethodHandle is a known object

Issue: #10618

### CODE REVIEW

**Comment by nbhuiyan** [comment]
@andrewcraik @liqunl Requesting review.

**Comment by andrewcraik** [comment]
@liqunl can you make a first pass please?

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp [hunk 1]
```diff
@@ -116,6 +116,80 @@ void J9::RecognizedCallTransformer::process_java_lang_StrictMath_and_Math_sqrt(T
...
+             aloadi <MethodHandle.type>
+               ==>aload <MethodHandle>
+...
+end Block_A
+
+*/
+void J9::RecognizedCallTransformer::process_java_lang_invoke_Invokers_checkExactType(TR::TreeTop* treetop, TR::Node* node)
+   {
+   TR::Node * methodHandleNode = node->getChild(0);
+   TR::Node * expectedTypeNode = node->getChild(1);
```
Comment (liqunl) [inline runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp:152]: It's better to use `getArgument`. The child index may not match argument index for some methods

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp [hunk 2]
```diff
@@ -116,6 +116,80 @@ void J9::RecognizedCallTransformer::process_java_lang_StrictMath_and_Math_sqrt(T
...
+   TR::SymbolReference *typeSymRef = comp()->getSymRefTab()->findOrFabricateShadowSymbol(comp()->getMethodSymbol(),
+                                                                                         TR::Symbol::Java_lang_invoke_MethodHandle_type,
+                                                                                         TR::Address,
+                                                                                         typeOffset,
+                                                                                         false,
+                                                                                         false,
+                                                                                         true,
+                                                                                         "java/lang/invoke/MethodHandle.type Ljava/lang/invoke/MethodType;");
+   TR::Node *handleTypeNode = TR::Node::createWithSymRef(comp()->il.opCodeForIndirectLoad(TR::Address), 1, 1, methodHandleNode, typeSymRef);
```
Comment (liqunl) [inline runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp:163]: Nodes created after ilgen should have originating bytecode node. Use the node create API that takes a node as the first argument to create the node.

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp [hunk 3]
```diff
@@ -116,6 +116,80 @@ void J9::RecognizedCallTransformer::process_java_lang_StrictMath_and_Math_sqrt(T
...
+   TR::Node *cmpEqNode = TR::Node::create(TR::acmpeq, 2, expectedTypeNode, handleTypeNode);
+   prepareToReplaceNode(node);
+   TR::Node::recreate(node, TR::ZEROCHK);
```
Comment (liqunl) [inline runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp:166]: The zerochk node should have MethodTypeCheck symref on it

File: runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp [hunk 4]
```diff
@@ -386,6 +460,12 @@ bool J9::RecognizedCallTransformer::isInlineable(TR::TreeTop* treetop)
       case TR::java_lang_StrictMath_sqrt:
       case TR::java_lang_Math_sqrt:
          return comp()->target().cpu.getSupportsHardwareSQRT();
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+      case TR::java_lang_invoke_Invokers_checkExactType:
+         return true;
+      case TR::java_lang_invoke_Invokers_checkCustomized:
+         return node->getChild(0)->getSymbolReference()->hasKnownObjectIndex();
```
Comment (liqunl) [inline runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp:467]: During inlining, the node may not have known object index. The known object info can come from inlined stack. Check if we're during inlining, and check the prex arg from the inline stack
```
TR_PrexArgInfo *argInfo = comp->getCurrentInlinedCallArgInfo();
   TR_PrexArgument *arg = argInfo->get(0);
   if (arg && arg->getKnownObjectIndex() != TR::KnownObjectTable::UNKNOWN)
      {   
```

**Comment by nbhuiyan** [comment]
@0xdaryl Requesting review

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdknext

**Comment by liqunl** [comment]
I found that RecongnizedCallTransformer is not a good place for MH opts, as transforming these recognized MH methods usually requires known object info, and we need final field folding to discover more known objects, also a way to propagate known object from autos to the call sites. I'm adding a new opt to do this, and all the transformations should go there.

**Comment by nbhuiyan** [comment]
Changing this PR to Draft state until we move these optimizations to a new opt.

**Comment by nbhuiyan** [comment]
@liqunl @0xdaryl requesting review.

File: runtime/compiler/optimizer/MethodHandleTransformer.cpp [hunk 5]
```diff
@@ -602,3 +608,87 @@ TR_MethodHandleTransformer::process_java_lang_invoke_MethodHandle_linkTo(TR::Tre
...
+             ==>aload <expected type>
+             aloadi <MethodHandle.type>
+               ==>aload <MethodHandle>
+...
+end Block_A
+
+*/
+void
+TR_MethodHandleTransformer::process_java_lang_invoke_Invokers_checkExactType(TR::TreeTop* tt, TR::Node* node)
+   {
```
Comment (liqunl) [inline runtime/compiler/optimizer/MethodHandleTransformer.cpp:642]: If the two arguments are known object, we can do the check at compile time and remove the call if the types match

**Comment by 0xdaryl** [comment]
@liqunl : please approve if you are satisfied with the latest changes.

File: runtime/compiler/optimizer/MethodHandleTransformer.cpp [hunk 6]
```diff
@@ -602,3 +609,123 @@ TR_MethodHandleTransformer::process_java_lang_invoke_MethodHandle_linkTo(TR::Tre
...
+      }
+   if (!performTransformation(comp(), "%sChanging checkExactType call node n%dn to ZEROCHK\n", optDetailString(), node->getGlobalIndex()))
+      return;
+   uint32_t typeOffset = fej9->getInstanceFieldOffsetIncludingHeader("Ljava/lang/invoke/MethodHandle;", "type", "Ljava/lang/invoke/MethodType;", comp()->getCurrentMethod());
+   auto typeSymRef = comp()->getSymRefTab()->findOrFabricateShadowSymbol(comp()->getMethodSymbol(),
+                                                                                         TR::Symbol::Java_lang_invoke_MethodHandle_type,
+                                                                                         TR::Address,
+                                                                                         typeOffset,
+                                                                                         false,
+                                                                                         false,
```
Comment (liqunl) [inline runtime/compiler/optimizer/MethodHandleTransformer.cpp:684]: `type` is a private field

**Comment by liqunl** [comment]
Other changes LGTM

**Review by liqunl — APPROVED** [review]

**Comment by 0xdaryl** [comment]
Jenkins test sanity xlinux,plinux,zlinux jdknext

Previous sanity testing passed.  I am going to just launch a shortened sanity run for the minor edits.

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**